dbms/include/DB/AggregateFunctions/AggregateFunctionGroupUniqArray.h

```diff
@@ -94,11 +95,142 @@ class AggregateFunctionGroupUniqArray
 		data_to.resize(old_size + size);
 
 		size_t i = 0;
-		for (typename State::Set::const_iterator it = set.begin(); it != set.end(); ++it, ++i)
+		for (auto it = set.begin(); it != set.end(); ++it, ++i)
 			data_to[old_size + i] = *it;
 	}
 };
 
+
+/// Generic implementation
+struct AggreagteFunctionGroupUniqArrayGenericData
+{
+	using Set = HashSetWithSavedHash<StringRef, StringRefHash, HashTableGrower<4>, HashTableAllocatorWithStackMemory<16>>;
+	Set value;
+};
+
+template <bool is_plain_column=false>
+class AggreagteFunctionGroupUniqArrayGeneric : public IUnaryAggregateFunction<AggreagteFunctionGroupUniqArrayGenericData, AggreagteFunctionGroupUniqArrayGeneric<is_plain_column>>
+{
+	mutable DataTypePtr input_data_type;
+
+	using State = AggreagteFunctionGroupUniqArrayGenericData;
+
+	static StringRef getSerialization(const IColumn & column, size_t row_num, Arena & arena);
+
+	static void deserializeAndInsert(StringRef str, IColumn & data_to);
+
+public:
+
+	String getName() const override { return "groupUniqArray"; }
+
+	void setArgument(const DataTypePtr & argument)
+	{
+		input_data_type = argument;
+	}
+
+	DataTypePtr getReturnType() const override
+	{
+		return std::make_shared<DataTypeArray>(input_data_type->clone());
+	}
+
+	void serialize(ConstAggregateDataPtr place, WriteBuffer & buf) const override
+	{
+		auto & set = this->data(place).value;
+		writeVarUInt(set.size(), buf);
+
+		for (auto & elem: set)
+		{
+			writeStringBinary(elem, buf);
+		}
+	}
+
+	void deserialize(AggregateDataPtr place, ReadBuffer & buf) const override
+	{
+		State::Set & set = this->data(place).value;
+		size_t size;
+		readVarUInt(size, buf);
+		//TODO: set.reserve(size);
+
+		std::string str_buf;
+		for (size_t i = 0; i < size; i++)
+		{
+			readStringBinary(str_buf, buf);
+			set.insert(StringRef(str_buf));
```

Comment:
Dangling StringRef. (Data for strings is not owned.)
Need to test for this method with distributed query processing.

It's worth to add `readStringBinaryInto` method, that reads directly into Arena, by analogy with `readStringInto` method.